From 3b0340a5b7137e25f337bcf4e010bd6c64cafb53 Mon Sep 17 00:00:00 2001
From: "pj.chen" <chen.pj@inventec.com>
Date: Tue, 1 Mar 2022 06:50:41 +0000
Subject: [PATCH 13/27] Bug 573 - [SW][Common] - Accumulate sensor initial
 commit

Sympton/Reason:
    Add accumulate sensor for the virtual sensor which needs accumulate several sensors.

Root Cause:
    N/A

Solution/Change:
    Accumulate sensor initial commit.

Entry Test:

{
	"Name": "acc_test",
	"SensorType": "temperature",
	"SensorMember1": "THERMAL_Q91",
	"SensorMember2": "THERMAL_Q93",
	"SensorMember3": "THERMAL_Q94",
	"SensorInfo": [
		{
			"Label": "acc_test",
			"SensorNum": "0x80",
			"EntityId": "0xa",
			"EntityInstance": "11"
		}
	],
	"Type": "AccumulateSensor"
}

sysadmin@transformers:~# ipmitool sdr elist
THERMAL_Q93      | 63h | ok  |  7.3 | 25 degrees C
THERMAL_Q91      | 64h | ok  |  7.4 | 25 degrees C
THERMAL_Q94      | 65h | ok  |  7.5 | 25 degrees C
THERMAL_Q92      | 66h | lcr |  7.6 | 0 degrees C
acc_test         | 80h | ok  | 10.11 | 75 degrees C

[comment for 2.12 upgrade]
Change code for threshold interfaces are moved in commit 5667808

---
 meson_options.txt                             |   1 +
 service_files/meson.build                     |   1 +
 ...z.openbmc_project.accumulatesensor.service |  12 +
 src/AccumulateSensor.cpp                      | 288 +++++++++++++++++
 src/AccumulateSensor.hpp                      |  47 +++
 src/AccumulateSensorMain.cpp                  | 304 ++++++++++++++++++
 src/meson.build                               |  16 +
 7 files changed, 669 insertions(+)
 create mode 100644 service_files/xyz.openbmc_project.accumulatesensor.service
 create mode 100644 src/AccumulateSensor.cpp
 create mode 100644 src/AccumulateSensor.hpp
 create mode 100644 src/AccumulateSensorMain.cpp

diff --git a/meson_options.txt b/meson_options.txt
index 9a75e7b..69ceb29 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -15,3 +15,4 @@ option('insecure-sensor-override', type : 'feature', value : 'disabled', descrip
 option('event', type: 'feature', value: 'enabled', description: 'Enable General event sensor.',)
 option('average', type: 'feature', value: 'enabled', description: 'Enable average sensor.',)
 option('invcfm', type: 'feature', value: 'enabled', description: 'Enable InvCfm sensor.',)
+option('accumulate', type: 'feature', value: 'enabled', description: 'Enable accumulate sensor.',)
diff --git a/service_files/meson.build b/service_files/meson.build
index 0e07444..34ff39a 100644
--- a/service_files/meson.build
+++ b/service_files/meson.build
@@ -13,6 +13,7 @@ unit_files = [
     ['event', 'xyz.openbmc_project.eventsensor.service'],
     ['average', 'xyz.openbmc_project.averagesensor.service'],
     ['invcfm', 'xyz.openbmc_project.invcfmsensor.service'],
+    ['accumulate', 'xyz.openbmc_project.accumulatesensor.service'],
 ]
 
 foreach tuple : unit_files
diff --git a/service_files/xyz.openbmc_project.accumulatesensor.service b/service_files/xyz.openbmc_project.accumulatesensor.service
new file mode 100644
index 0000000..f918f85
--- /dev/null
+++ b/service_files/xyz.openbmc_project.accumulatesensor.service
@@ -0,0 +1,12 @@
+[Unit]
+Description=Inventec Accumulate Sensor
+Requires=xyz.openbmc_project.EntityManager.service
+After=xyz.openbmc_project.EntityManager.service
+
+[Service]
+Restart=always
+RestartSec=5
+ExecStart=/usr/bin/accumulatesensor
+
+[Install]
+WantedBy=multi-user.target
diff --git a/src/AccumulateSensor.cpp b/src/AccumulateSensor.cpp
new file mode 100644
index 0000000..36a628c
--- /dev/null
+++ b/src/AccumulateSensor.cpp
@@ -0,0 +1,288 @@
+/*
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+/ You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "AccumulateSensor.hpp"
+
+#include <unistd.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/asio/read_until.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <sdbusplus/asio/connection.hpp>
+#include <sdbusplus/asio/object_server.hpp>
+
+#include <iostream>
+#include <istream>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+static constexpr unsigned int sensorPollMs = 500;
+
+AccumulateSensor::AccumulateSensor(
+    const std::string& objectType,
+    sdbusplus::asio::object_server& objectServer,
+    std::shared_ptr<sdbusplus::asio::connection>& conn,
+    boost::asio::io_context& io,
+    const std::string& sensorName,
+    const std::string& sensorType,
+    const std::string& eventType,
+    const double max, const double min,
+    std::vector<std::string>&& sensorMembers,
+    std::vector<thresholds::Threshold>&& _thresholds,
+    std::vector<ipmidata::IpmiConfig>&& _ipmiinfo,
+    const std::string& sensorConfiguration) :
+    Sensor(boost::replace_all_copy(sensorName, " ", "_"),
+           std::move(_thresholds), sensorConfiguration, objectType, false, false,
+           max, min, conn),
+    std::enable_shared_from_this<AccumulateSensor>(), objServer(objectServer),
+    waitTimer(io)
+{
+    std::string path;
+    std::string sensorMemberPathPrefix;
+
+    if(eventType == "")
+    {
+        path = "/xyz/openbmc_project/sensors/" + sensorType + "/" + sensorName;
+        sensorMemberPathPrefix = "/xyz/openbmc_project/sensors/" + sensorType;
+    }
+    else
+    {
+        path = "/xyz/openbmc_project/sensors/"+ eventType +
+            "/" + sensorType + "/" + sensorName;
+
+        sensorMemberPathPrefix = "/xyz/openbmc_project/sensors/"+ eventType +
+            "/" + sensorType;
+    }
+
+    for (auto& sensorMember : sensorMembers)
+    {
+        sensorMemberPaths.emplace_back(
+            sensorMemberPathPrefix + "/" + sensorMember);
+    }
+
+    sensorInterface = objectServer.add_interface(
+        path,
+        "xyz.openbmc_project.Sensor.Value");
+
+    ipmiInterface = objectServer.add_interface(
+        path,
+        "xyz.openbmc_project.Sensor.IpmiSensor");
+
+    if (_ipmiinfo.empty())
+    {
+        // Default
+        ipmiInterface->register_property("sensorNumber", 0);
+        ipmiInterface->register_property("entityID", 0);
+        ipmiInterface->register_property("entityInstance", 0);
+    }
+    else
+    {
+        for (auto& ipmi : _ipmiinfo)
+        {
+            ipmiInterface->register_property("sensorNumber", ipmi.number);
+            ipmiInterface->register_property("entityID", ipmi.entityId);
+            ipmiInterface->register_property("entityInstance", ipmi.entityIns);
+        }
+    }
+
+    for (const auto& threshold : thresholds)
+    {
+        std::string interface =
+            thresholds::getInterface(threshold.level);
+        thresholdInterfaces[static_cast<size_t>(threshold.level)] =
+            objectServer.add_interface(path, interface);
+    }
+    association = objectServer.add_interface(
+        path,
+        association::interface);
+
+    const char* units;
+
+    if(sensorType == "temperature")
+    {
+        units = sensor_paths::unitDegreesC;
+    }
+    else if(sensorType == "fan_tach")
+    {
+        units = sensor_paths::unitRPMs;
+    }
+    else if(sensorType == "voltage")
+    {
+        units = sensor_paths::unitVolts;
+    }
+    else if(sensorType == "current")
+    {
+        units = sensor_paths::unitAmperes;
+    }
+    else if (sensorType == "power")
+    {
+        units = sensor_paths::unitWatts;
+    }
+    else if(sensorType == "energy")
+    {
+        units = sensor_paths::unitJoules;
+    }
+    else if(sensorType == "fan_pwm")
+    {
+        units = sensor_paths::unitPercent;
+    }
+    else if(sensorType == "cfm")
+    {
+        units = sensor_paths::unitCFM;
+    }
+    else
+    {
+        /* Not found, set to default unitDegreesC*/
+        units = sensor_paths::unitDegreesC;
+    }
+
+    setInitialProperties(units);
+}
+
+AccumulateSensor::~AccumulateSensor()
+{
+    // close the input dev to cancel async operations
+    for (const auto& iface : thresholdInterfaces)
+    {
+        objServer.remove_interface(iface);
+    }
+    objServer.remove_interface(sensorInterface);
+    objServer.remove_interface(ipmiInterface);
+    objServer.remove_interface(association);
+}
+
+void AccumulateSensor::getSensorReading(std::string& path)
+{
+    std::array<const char*, 4> interfaces = {
+            "xyz.openbmc_project.Sensor.Value",
+            "xyz.openbmc_project.Sensor.Threshold.Warning",
+            "xyz.openbmc_project.Sensor.Threshold.Critical",
+            "xyz.openbmc_project.Sensor.IpmiSensor"};
+
+    dbusConnection->async_method_call(
+        [&](const boost::system::error_code ec,
+            const std::vector<std::pair<std::string, std::vector<std::string>>>& objects) {
+            if (ec)
+            {
+                updatedSensorNumber++;
+            }
+            else
+            {
+                dbusConnection->async_method_call(
+                    [&](boost::system::error_code ec,
+                        const std::variant<double>& value) {
+                        updatedSensorNumber++;
+                        if (ec)
+                        {
+                            // sensor not ready yet
+                            std::fprintf(stderr, "Error getting value %s\n", path.c_str());
+                        }
+                        else
+                        {
+                            sensorValues.emplace_back(
+                                std::visit(VariantToDoubleVisitor(), value));
+                        }
+                        /* Check update after get values */
+                        if(updatedSensorNumber == sensorMemberPaths.size())
+                        {
+                            updateReading();
+                        }
+                        return;
+                    },
+                    objects.begin()->first, path,
+                    properties::interface, properties::get, sensorValueInterface,
+                    "Value");
+            }
+            /* Check update after GetOject fail */
+            if(updatedSensorNumber == sensorMemberPaths.size())
+            {
+                updateReading();
+            }
+            return;
+        },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetObject",
+        path, interfaces);
+
+    return;
+}
+
+void AccumulateSensor::setupRead(void)
+{
+    /* Clear the sensor number of updated, wait for new update.*/
+    updatedSensorNumber = 0;
+    sensorValues.clear();
+
+    for (auto& sensorMemberPath : sensorMemberPaths)
+    {
+        getSensorReading(sensorMemberPath);
+    }
+}
+
+void AccumulateSensor::checkUpdate(void)
+{
+
+    if(updatedSensorNumber == sensorMemberPaths.size())
+    {
+        updateReading();
+    }
+}
+
+void AccumulateSensor::updateReading(void)
+{
+    double sum = 0;
+    /* update to dbus */
+    if( !sensorValues.empty())
+    {
+        for (auto& sensorValue : sensorValues)
+        {
+            sum += sensorValue;
+        }
+        updateValue(sum);
+    }
+
+    /* Set next read */
+    waitTimer.expires_from_now(boost::posix_time::milliseconds(sensorPollMs));
+    std::weak_ptr<AccumulateSensor> weakRef = weak_from_this();
+    waitTimer.async_wait([weakRef](const boost::system::error_code& ec) {
+        std::shared_ptr<AccumulateSensor> self = weakRef.lock();
+        if (ec == boost::asio::error::operation_aborted)
+        {
+            if (self)
+            {
+                std::cerr << "Accumulate sensor " << self->name << "\n";
+            }
+            else
+            {
+                std::cerr << "Accumulate sensor read cancelled, no self\n";
+            }
+            return; // we're being canceled
+        }
+        if (self)
+        {
+            self->setupRead();
+        }
+    });
+}
+
+void AccumulateSensor::checkThresholds(void)
+{
+    thresholds::checkThresholds(this);
+}
diff --git a/src/AccumulateSensor.hpp b/src/AccumulateSensor.hpp
new file mode 100644
index 0000000..cd0017f
--- /dev/null
+++ b/src/AccumulateSensor.hpp
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <Thresholds.hpp>
+#include <Utils.hpp>
+#include <sensor.hpp>
+#include <IpmiInfo.hpp>
+
+#include <boost/asio/deadline_timer.hpp>
+#include <boost/asio/streambuf.hpp>
+#include <sdbusplus/asio/object_server.hpp>
+
+#include <string>
+#include <vector>
+
+class AccumulateSensor :
+    public Sensor,
+    public std::enable_shared_from_this<AccumulateSensor>
+{
+  public:
+    AccumulateSensor( const std::string& objectType,
+                    sdbusplus::asio::object_server& objectServer,
+                    std::shared_ptr<sdbusplus::asio::connection>& conn,
+                    boost::asio::io_context& io,
+                    const std::string& sensorName,
+                    const std::string& sensorType,
+                    const std::string& eventType,
+                    const double max, const double min,
+                    std::vector<std::string>&& sensorMembers,
+                    std::vector<thresholds::Threshold>&& thresholds,
+                    std::vector<ipmidata::IpmiConfig>&& ipmiinfo,
+                    const std::string& sensorConfiguration);
+    ~AccumulateSensor();
+    void setupRead(void);
+
+  private:
+    sdbusplus::asio::object_server& objServer;
+    boost::asio::deadline_timer waitTimer;
+    size_t errCount;
+    std::vector<std::string> sensorMemberPaths;
+    std::vector<double> sensorValues;
+    unsigned int updatedSensorNumber;
+
+    void getSensorReading(std::string& path);
+    void checkUpdate(void);
+    void updateReading(void);
+    void checkThresholds(void) override;
+};
diff --git a/src/AccumulateSensorMain.cpp b/src/AccumulateSensorMain.cpp
new file mode 100644
index 0000000..67862e7
--- /dev/null
+++ b/src/AccumulateSensorMain.cpp
@@ -0,0 +1,304 @@
+/*
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "AccumulateSensor.hpp"
+#include "Utils.hpp"
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/container/flat_map.hpp>
+#include <boost/container/flat_set.hpp>
+#include <sdbusplus/asio/connection.hpp>
+#include <sdbusplus/asio/object_server.hpp>
+#include <sdbusplus/bus/match.hpp>
+
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <functional>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+static constexpr bool DEBUG = false;
+
+namespace fs = std::filesystem;
+static constexpr std::array<const char*, 1> accumulateTypes = {
+    "xyz.openbmc_project.Configuration.AccumulateSensor"};
+
+void createSensors(
+    boost::asio::io_context& io, sdbusplus::asio::object_server& objectServer,
+    boost::container::flat_map<std::string, std::shared_ptr<AccumulateSensor>>&
+        sensors,
+    std::shared_ptr<sdbusplus::asio::connection>& dbusConnection,
+    const std::shared_ptr<boost::container::flat_set<std::string>>&
+        sensorsChanged)
+{
+    auto getter = std::make_shared<GetSensorConfiguration>(
+        dbusConnection,
+        std::move([&io, &objectServer, &sensors, &dbusConnection,
+                   sensorsChanged](
+                      const ManagedObjectType& sensorConfigurations) {
+            bool firstScan = sensorsChanged == nullptr;
+
+            const SensorData* sensorData = nullptr;
+            const std::string* interfacePath = nullptr;
+            const char* accumulateType = nullptr;
+            const SensorBaseConfiguration* baseConfiguration = nullptr;
+            const SensorBaseConfigMap* baseConfigMap = nullptr;
+
+            for (const std::pair<sdbusplus::message::object_path,
+                                    SensorData>& sensorConf : sensorConfigurations)
+            {
+                /* Reset variables*/
+                sensorData = nullptr;
+                interfacePath = nullptr;
+                accumulateType = nullptr;
+                baseConfiguration = nullptr;
+
+                sensorData = &(sensorConf.second);
+                for (const char* type : accumulateTypes)
+                {
+                    auto sensorBase = sensorData->find(type);
+                    if (sensorBase != sensorData->end())
+                    {
+                        baseConfiguration = &(*sensorBase);
+                        accumulateType = type;
+                        break;
+                    }
+                }
+                if (baseConfiguration == nullptr)
+                {
+                    continue;
+                }
+
+                baseConfigMap = &baseConfiguration->second;
+                interfacePath = &(sensorConf.first.str);
+
+                /* Find Sensor Name*/
+                auto findSensorName = baseConfigMap->find("Name");
+                if (findSensorName == baseConfigMap->end())
+                {
+                    std::cerr << "could not determine configuration name for "
+                                << *interfacePath << "\n";
+                    continue;
+                }
+                std::string sensorName =
+                    std::get<std::string>(findSensorName->second);
+                // on rescans, only update sensors we were signaled by
+                auto findSensor = sensors.find(sensorName);
+                if (!firstScan && findSensor != sensors.end())
+                {
+                    bool found = false;
+                    for (auto it = sensorsChanged->begin();
+                            it != sensorsChanged->end(); it++)
+                    {
+                        if (boost::ends_with(*it, findSensor->second->name))
+                        {
+                            sensorsChanged->erase(it);
+                            findSensor->second = nullptr;
+                            found = true;
+                            break;
+                        }
+                    }
+                    if (!found)
+                    {
+                        continue;
+                    }
+                }
+
+                /* Find Sensor type */
+                auto findSensorType = baseConfigMap->find("SensorType");
+                if (findSensorType == baseConfigMap->end())
+                {
+                    std::cerr << "could not determine configuration SensorType for "
+                                << *interfacePath << "\n";
+                    continue;
+                }
+                std::string sensorType =
+                    std::get<std::string>(findSensorType->second);
+
+                /* Find Event type */
+                std::string eventType;
+                auto findEventType = baseConfigMap->find("EventType");
+                if (findEventType == baseConfigMap->end())
+                {
+                    std::cerr << "could not determine configuration EventType for "
+                                << *interfacePath << "\n";
+
+                    eventType = "";
+                }
+                else
+                {
+                    eventType =
+                        std::get<std::string>(findEventType->second);
+                }
+
+                /* Find Max*/
+                double max;
+                auto findMax = baseConfigMap->find("Max");
+                if (findMax != baseConfigMap->end())
+                {
+                    try
+                    {
+                        max = std::visit(VariantToDoubleVisitor(), findMax->second);
+                    }
+                    catch (std::invalid_argument&)
+                    {
+                        std::cerr << "Unable to parse Max\n";
+                        continue;
+                    }
+                }
+
+                /* Find Min*/
+                double min;
+                auto findMin = baseConfigMap->find("Min");
+                if (findMin != baseConfigMap->end())
+                {
+                    try
+                    {
+                       min = std::visit(VariantToDoubleVisitor(), findMin->second);
+                    }
+                    catch (std::invalid_argument&)
+                    {
+                        std::cerr << "Unable to parse Min\n";
+                        continue;
+                    }
+                }
+
+                auto& sensor = sensors[sensorName];
+                sensor = nullptr;
+
+                std::vector<thresholds::Threshold> sensorThresholds;
+                if (!parseThresholdsFromConfig(*sensorData, sensorThresholds))
+                {
+                    std::cerr << "error populating thresholds for "
+                                << sensorName << "\n";
+                }
+
+                std::vector<ipmidata::IpmiConfig> sensorIpmiConfig;
+                if (!parseIpmiDataFromConfig(*sensorData, sensorIpmiConfig, &sensorName))
+                {
+                    std::cerr << "error populating ipmi for "
+                                << sensorName << "\n";
+                }
+
+                /* Find member sensors */
+                int i = 0;
+                std::vector<std::string> sensorMembers;
+                while (true)
+                {
+                    ++i;
+                    auto findSensorMember =
+                        baseConfigMap->find("SensorMember" + std::to_string(i));
+                    if (findSensorMember == baseConfigMap->end())
+                    {
+                        break;
+                    }
+                    sensorMembers.emplace_back(
+                        std::get<std::string>(findSensorMember->second));
+                }
+
+                if (DEBUG)
+                {
+                    std::cerr <<"SensorName: " << sensorName << '\n'
+                        << "AccumulateType: " << accumulateType << '\n'
+                        << "SensorType: " << sensorType << '\n'
+                        << "EventType: " << eventType << '\n'
+                        << "Max: " << max << '\n'
+                        << "Min: " << min << '\n'
+                        << "Interface:" << *interfacePath << '\n';
+
+                    for (auto& sensorMember : sensorMembers)
+                    {
+                        std::cerr << "sensorMember " << sensorMember << '\n';
+                    }
+                }
+
+                sensor = std::make_shared<AccumulateSensor>(
+                    accumulateType, objectServer, dbusConnection,
+                    io, sensorName, sensorType, eventType,
+                    max, min,
+                    std::move(sensorMembers),
+                    std::move(sensorThresholds),
+                    std::move(sensorIpmiConfig),
+                    *interfacePath);
+                sensor->setupRead();
+            }
+        }));
+    getter->getConfiguration(
+        std::vector<std::string>(accumulateTypes.begin(), accumulateTypes.end()));
+}
+
+int main()
+{
+    boost::asio::io_context io;
+    auto systemBus = std::make_shared<sdbusplus::asio::connection>(io);
+    systemBus->request_name("xyz.openbmc_project.AccumulateSensor");
+    sdbusplus::asio::object_server objectServer(systemBus);
+    boost::container::flat_map<std::string, std::shared_ptr<AccumulateSensor>>
+        sensors;
+    std::vector<std::unique_ptr<sdbusplus::bus::match::match>> matches;
+    auto sensorsChanged =
+        std::make_shared<boost::container::flat_set<std::string>>();
+
+    boost::asio::post(io, [&]() {
+        createSensors(io, objectServer, sensors, systemBus, nullptr);
+    });
+
+    boost::asio::deadline_timer filterTimer(io);
+    std::function<void(sdbusplus::message::message&)> eventHandler =
+        [&](sdbusplus::message::message& message) {
+            if (message.is_method_error())
+            {
+                std::cerr << "callback method error\n";
+                return;
+            }
+            sensorsChanged->insert(message.get_path());
+            // this implicitly cancels the timer
+            filterTimer.expires_from_now(boost::posix_time::seconds(1));
+
+            filterTimer.async_wait([&](const boost::system::error_code& ec) {
+                if (ec == boost::asio::error::operation_aborted)
+                {
+                    /* we were canceled*/
+                    return;
+                }
+                else if (ec)
+                {
+                    std::cerr << "timer error\n";
+                    return;
+                }
+                createSensors(io, objectServer, sensors, systemBus,
+                              sensorsChanged);
+            });
+        };
+
+    for (const char* type : accumulateTypes)
+    {
+        auto match = std::make_unique<sdbusplus::bus::match::match>(
+            static_cast<sdbusplus::bus::bus&>(*systemBus),
+            "type='signal',member='PropertiesChanged',path_namespace='" +
+                std::string(inventoryPath) + "',arg0namespace='" + type + "'",
+            eventHandler);
+        matches.emplace_back(std::move(match));
+    }
+    io.run();
+}
diff --git a/src/meson.build b/src/meson.build
index 5f49e71..22c0e21 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -290,3 +290,19 @@ if get_option('invcfm').enabled()
         install: true,
     )
 endif
+
+if get_option('accumulate').enabled()
+    executable(
+        'accumulatesensor',
+        'AccumulateSensor.cpp',
+        'AccumulateSensorMain.cpp',
+        'IpmiInfo.cpp',
+        dependencies: [
+            default_deps,
+            thresholds_dep,
+            utils_dep,
+        ],
+        cpp_args: uring_args,
+        install: true,
+    )
+endif
-- 
2.25.1